src/dpl/src/Graphics.cpp

```diff
@@ -0,0 +1,166 @@
+                   bool displacement,
+                   float min_displacement,
+                   const dbInst* debug_instance) :
+  dp_(dp),
+  debug_instance_(debug_instance),
+  block_(nullptr),
+  displacement_(displacement),
+  min_displacement_(min_displacement)
+{
+  gui::Gui::get()->registerRenderer(this);
+}
+
+void
+Graphics::startPlacement(dbBlock *block)
+{
+  block_ = block;
+}
+
+void
+Graphics::placeInstance(dbInst *instance)
+{
+  if (!instance || instance != debug_instance_) {
+    return;
+  }
+
+  auto gui = gui::Gui::get();
+
+  auto selected = gui->makeSelected(instance);
+  gui->setSelected(selected);
+  gui->redraw();
+  gui->pause();
+}
+
+void
+Graphics::binSearch(const Cell *cell,
+                    int xl,
+                    int yl,
+                    int xh,
+                    int yh)
+{
+  if (!debug_instance_ || cell->db_inst_ != debug_instance_) {
+    return;
+  }
+  Rect core = dp_->getCore();
+  int xl_dbu = core.xMin() + xl * dp_->getSiteWidth();
+  int yl_dbu = core.yMin() + yl * dp_->getRowHeight();
+  int xh_dbu = core.xMin() + xh * dp_->getSiteWidth();
+  int yh_dbu = core.yMin() + yh * dp_->getRowHeight();
+  searched_.push_back(Rect(xl_dbu, yl_dbu, xh_dbu, yh_dbu));
+}
+
+void
+Graphics::endPlacement()
+{
+  auto gui = gui::Gui::get();
+  gui->redraw();
+  gui->pause();
+}
+
+void
+Graphics::drawObjects(gui::Painter &painter)
+{
+  if (!block_) {
+    return;
+  }
+
+  // Compare the squared distances to save calling sqrt
+  float min_length = min_displacement_ * dp_->getRowHeight();
+  min_length *= min_length;
+
+  auto core = ord::getCore(block_);
+
+  for (auto cell : dp_->getCells()) {
+    if (!cell.is_placed_) {
+      continue;
+    }
+
+    int lx = core.xMin() + cell.x_;
+    int ly = core.yMin() + cell.y_;
+
+    auto color = cell.db_inst_ ? gui::Painter::gray : gui::Painter::red;
+    painter.setPen(color);
+    painter.setBrush(color);
+    painter.drawRect(Rect(lx, ly, lx + cell.width_, ly + cell.height_));
+
+    if (!cell.db_inst_) {
+      continue;
+    }
+
+    dbBox *bbox = cell.db_inst_->getBBox();
+    Point initial_location(bbox->xMin(), bbox->yMin());
+    Point final_location(lx, ly);
+    float len = Point::squaredDistance(initial_location,
+                                       final_location);
+    if (len < min_length) {
+      continue;
+    }
+
+    painter.setPen(gui::Painter::yellow, /* cosmetic */ true);
+    painter.drawLine(initial_location.x(),
+                     initial_location.y(),
+                     final_location.x(),
+                     final_location.y());
+    painter.drawCircle(final_location.x(),
+                       final_location.y(),
+                       100);
+  }
+
+  auto color = gui::Painter::cyan;
+  painter.setPen(color);
+  painter.setBrush(color);
+  for (auto& rect : searched_) {
+    painter.drawRect(rect);
+  }
+}
+
+/* static */
+bool
+Graphics::guiActive()
+{
+  return gui::Gui::get() != nullptr;
```

Comment:
this will always be true, you should use the `gui::Gui::enabled()` instead

Reply:
thanks I started this before your change and missed updating it.